Adjust result type of product Tensor*scalar

bangerth opened the pull request.

bangerth commented:
By way of explanation:

```
+  <li> Changed: If you take the product of a Tensor and a scalar number,
+  you previously got a Tensor back that stored its elements in the same
+  data type as the original tensor. This leads to problems if you
+  multiply a <code>Tensor@<1,dim,double@></code> by a
+  <code>std::complex@<double@></code> because the result clearly needs
+  to store its elements as complex numbers, rather than as double
+  variables. This is now changed: The result of the product of a Tensor
+  and a scalar number is now a Tensor that stores its elements in a data
+  type appropriate for this product.
```

kronbichler commented on include/deal.II/base/template_constraints.h:
```diff
@@ -297,6 +300,236 @@ struct types_are_equal<T,T>
+ * This class provides a local typedef @p type that is equal to the
+ * template argument but only if the template argument corresponds to
+ * a scalar type (i.e., one of the floating point types, signed or unsigned
+ * integer, or a complex number. If the template type @p T is not a scalar,
```
Closing bracket missing

kronbichler commented on source/grid/grid_generator.cc:
```diff
@@ -777,20 +777,20 @@ namespace GridGenerator
       case 1:
         for (unsigned int x=0; x<=n_subdivisions[0]; ++x)
-          points.push_back (Point<dim> (x*delta[0]));
+          points.push_back (Point<dim> (double(x)*delta[0]));
```
Just to help me understand the chain of EnableIfScalar and ProductType: You need to cast x to double here because ProductType of unsigned int and double does not exist, right? Because unsigned int passes the EnableIfScalar declaration.

bangerth replied:
That's not actually the point. I cast because the compiler reports an ambiguity between available operators `double*Point` and `int*Tensor`, both of which require an implicit cast. The latter is the new operator introduced in this patch. I'm just choosing which of the two operators I want the compiler to use.

Given the selection, I can get rid of the (pre-existing) explicit cast which I will do right away :-)

kronbichler commented:
Looks good.

Given how often we use operations on tensor, I imagine that this might increase compile times of the library by a few percent... :-)

bangerth commented:
Yes, I agree that this may lead to longer compile times. I don't see much of an alternative, though, if we want to support operations such as `complex*Tensor` when forming the gradients at quadrature points like in `\grad u_h(xq) = \sum_j U_j \grad\varphi_j(xq)` if the solution vector is complex.

I also think it's sort of the right thing to do.

I've fixed the missing parenthesis and updated the other places.

kronbichler commented:
I definitely agree that this is the right way to go. I was just realizing it myself.
The changes look good. I will merge.

tjhei commented:
I am a bit late to this party, but:
- Did you check that the compiler outputs similar machine code that gives the same performance for the double*Tensor case?
- My vote would be to only allow multiplication with double or complex because multiplication with a bool, int, or something like that is probably unintended and shouldn't compile (or maybe convert to double implicitly).

bangerth commented:
In the place where I do the product, I convert both the scalar multiplier and the elements of the tensor to the result type, and then do the multiplication. So if you have
```
  double a;
  Tensor<1,dim,double> t;
  a*t;   // or t*a
```
then the product happens in double precision. The same if `a` or `t` was using `float`. If you have
```
  std::complex<float> a;
  Tensor<1,double> t;
  a*t;   // or t*a
```
then both `a` and the elements of `t` are first converted to `complex<double>` and then multiplied.

I'm not sure this answers your first question. Feel free to rephrase if necessary.

For your second question, we do need to allow multiplication with an integer because people do write `2*t` and I think that's reasonable. For `bool` it's probably true that that's unintended. It should currently fail because `EnableIsScalar<bool>` is not declared.

tjhei commented:
> I'm not sure this answers your first question.

No. My question is: Are you sure (did you check?) that the new code is not slower than the old one?

> we do need to allow multiplication with an integer because people do write ```2*t``` and I think that's reasonable. 

I think this is likely a user error and she probably meant ```2.0*t```.

bangerth commented:
> No. My question is: Are you sure (did you check?) that the new code is
> not slower than the old one?

I did not check. Concretely, do you want me to check whether 
`double*Tensor<1,dim,double>` produces the same code as before?


>     we do need to allow multiplication with an integer because people do
>     write |2*t| and I think that's reasonable.
>
> I think this is likely a user error and she probably meant |2.0*t|.

I think that's debatable. It worked before and people write it all the 
time. We also write `2*3.141` and it works. I think allowing this is 
reasonable.

bangerth commented:
OK, part 1 of the answer: The multiplication with a number of the same type as stored in the tensor remains special cased and unchanged:
```
/**
 * Multiplication of a tensor of rank 1 with a scalar Number from the right.
 *
 * @relates Tensor<1,dim,Number>
 */
template <int dim, typename Number>
inline
Tensor<1,dim,Number>
operator * (const Tensor<1,dim,Number> &t,
            const Number                factor)
{
  Tensor<1,dim,Number> tt (false);
  for (unsigned int d=0; d<dim; ++d)
    tt[d] = t[d] * factor;
  return tt;
}



/**
 * Multiplication of a tensor of rank 1 with a scalar Number from the left.
 *
 * @relates Tensor<1,dim,Number>
 */
template <int dim, typename Number>
inline
Tensor<1,dim,Number>
operator * (const Number                factor,
            const Tensor<1,dim,Number> &t)
{
  Tensor<1,dim,Number> tt (false);
  for (unsigned int d=0; d<dim; ++d)
    tt[d] = t[d] * factor;
  return tt;
}
```
So there is definitely no speed loss for the very common case of multiplying a "norm" tensor with a double.

bangerth commented:
Second answer: With these two functions (the second being a renamed `operator*`)
```
template <int dim>
Tensor<1,dim> old_multiply (const Tensor<1,dim> &t,
			    const double d)
{
  Tensor<1,dim> tt;
  for (unsigned int i=0; i<dim; ++i)
    tt[i] = d * t[i];
  return tt;
}

template <int dim, typename Number, typename OtherNumber>
Tensor<1,dim,typename ProductType<Number,typename EnableIfScalar<OtherNumber>::type>::type>
new_multiply (const Tensor<1,dim,Number> &t,
	      const OtherNumber           factor)
{
  // form the product. we have to convert the two factors into the final
  // type via explicit casts because, for awkward reasons, the C++
  // standard committee saw it fit to not define an
  //   operator*(float,std::complex<double>)
  // (as well as with switched arguments and double<->float).
  typedef typename ProductType<Number,OtherNumber>::type product_type;
  Tensor<1,dim,product_type> tt (false);
  for (unsigned int d=0; d<dim; ++d)
    tt[d] = product_type(t[d]) * product_type(factor);
  return tt;
}
```
I get disassembly as follows:
```
Dump of assembler code for function old_multiply<3>(dealii::Tensor<1, 3, double> const&, double):
   0x00000000004141ef <+0>:	push   %rbp
   0x00000000004141f0 <+1>:	mov    %rsp,%rbp
   0x00000000004141f3 <+4>:	push   %rbx
   0x00000000004141f4 <+5>:	sub    $0x38,%rsp
   0x00000000004141f8 <+9>:	mov    %rdi,-0x28(%rbp)
   0x00000000004141fc <+13>:	mov    %rsi,-0x30(%rbp)
   0x0000000000414200 <+17>:	movsd  %xmm0,-0x38(%rbp)
=> 0x0000000000414205 <+22>:	mov    -0x28(%rbp),%rax
   0x0000000000414209 <+26>:	mov    $0x1,%esi
   0x000000000041420e <+31>:	mov    %rax,%rdi
   0x0000000000414211 <+34>:	callq  0x4141b2 <dealii::Tensor<1, 3, double>::Tensor(bool)>
   0x0000000000414216 <+39>:	movl   $0x0,-0x14(%rbp)
   0x000000000041421d <+46>:	jmp    0x414251 <old_multiply<3>(dealii::Tensor<1, 3, double> const&, double)+98>
   0x000000000041421f <+48>:	mov    -0x14(%rbp),%edx
   0x0000000000414222 <+51>:	mov    -0x28(%rbp),%rax
   0x0000000000414226 <+55>:	mov    %edx,%esi
   0x0000000000414228 <+57>:	mov    %rax,%rdi
   0x000000000041422b <+60>:	callq  0x414928 <dealii::Tensor<1, 3, double>::operator[](unsigned int)>
   0x0000000000414230 <+65>:	mov    %rax,%rbx
   0x0000000000414233 <+68>:	mov    -0x14(%rbp),%edx
   0x0000000000414236 <+71>:	mov    -0x30(%rbp),%rax
   0x000000000041423a <+75>:	mov    %edx,%esi
   0x000000000041423c <+77>:	mov    %rax,%rdi
   0x000000000041423f <+80>:	callq  0x41485a <dealii::Tensor<1, 3, double>::operator[](unsigned int) const>
   0x0000000000414244 <+85>:	mulsd  -0x38(%rbp),%xmm0
   0x0000000000414249 <+90>:	movsd  %xmm0,(%rbx)
   0x000000000041424d <+94>:	addl   $0x1,-0x14(%rbp)
   0x0000000000414251 <+98>:	cmpl   $0x2,-0x14(%rbp)
   0x0000000000414255 <+102>:	setbe  %al
   0x0000000000414258 <+105>:	test   %al,%al
   0x000000000041425a <+107>:	jne    0x41421f <old_multiply<3>(dealii::Tensor<1, 3, double> const&, double)+48>
   0x000000000041425c <+109>:	nop
   0x000000000041425d <+110>:	mov    -0x28(%rbp),%rax
   0x0000000000414261 <+114>:	add    $0x38,%rsp
   0x0000000000414265 <+118>:	pop    %rbx
   0x0000000000414266 <+119>:	pop    %rbp
   0x0000000000414267 <+120>:	retq
```
In other words, they are identical.
and
```
Dump of assembler code for function new_multiply<3, double, double>(dealii::Tensor<1, 3, double> const&, double):
   0x0000000000414268 <+0>:	push   %rbp
   0x0000000000414269 <+1>:	mov    %rsp,%rbp
   0x000000000041426c <+4>:	push   %rbx
   0x000000000041426d <+5>:	sub    $0x38,%rsp
   0x0000000000414271 <+9>:	mov    %rdi,-0x28(%rbp)
   0x0000000000414275 <+13>:	mov    %rsi,-0x30(%rbp)
   0x0000000000414279 <+17>:	movsd  %xmm0,-0x38(%rbp)
=> 0x000000000041427e <+22>:	mov    -0x28(%rbp),%rax
   0x0000000000414282 <+26>:	mov    $0x0,%esi
   0x0000000000414287 <+31>:	mov    %rax,%rdi
   0x000000000041428a <+34>:	callq  0x4141b2 <dealii::Tensor<1, 3, double>::Tensor(bool)>
   0x000000000041428f <+39>:	movl   $0x0,-0x14(%rbp)
   0x0000000000414296 <+46>:	jmp    0x4142ca <new_multiply<3, double, double>(dealii::Tensor<1, 3, double> const&, double)+98>
   0x0000000000414298 <+48>:	mov    -0x14(%rbp),%edx
   0x000000000041429b <+51>:	mov    -0x28(%rbp),%rax
   0x000000000041429f <+55>:	mov    %edx,%esi
   0x00000000004142a1 <+57>:	mov    %rax,%rdi
   0x00000000004142a4 <+60>:	callq  0x414928 <dealii::Tensor<1, 3, double>::operator[](unsigned int)>
   0x00000000004142a9 <+65>:	mov    %rax,%rbx
   0x00000000004142ac <+68>:	mov    -0x14(%rbp),%edx
   0x00000000004142af <+71>:	mov    -0x30(%rbp),%rax
   0x00000000004142b3 <+75>:	mov    %edx,%esi
   0x00000000004142b5 <+77>:	mov    %rax,%rdi
   0x00000000004142b8 <+80>:	callq  0x41485a <dealii::Tensor<1, 3, double>::operator[](unsigned int) const>
   0x00000000004142bd <+85>:	mulsd  -0x38(%rbp),%xmm0
   0x00000000004142c2 <+90>:	movsd  %xmm0,(%rbx)
   0x00000000004142c6 <+94>:	addl   $0x1,-0x14(%rbp)
   0x00000000004142ca <+98>:	cmpl   $0x2,-0x14(%rbp)
   0x00000000004142ce <+102>:	setbe  %al
   0x00000000004142d1 <+105>:	test   %al,%al
   0x00000000004142d3 <+107>:	jne    0x414298 <new_multiply<3, double, double>(dealii::Tensor<1, 3, double> const&, double)+48>
   0x00000000004142d5 <+109>:	nop
   0x00000000004142d6 <+110>:	mov    -0x28(%rbp),%rax
   0x00000000004142da <+114>:	add    $0x38,%rsp
   0x00000000004142de <+118>:	pop    %rbx
   0x00000000004142df <+119>:	pop    %rbp
   0x00000000004142e0 <+120>:	retq
```

tjhei commented:
>>> we do need to allow multiplication with an integer because people do
>>>     write |2*t| and I think that's reasonable.
>>
>> I think this is likely a user error and she probably meant |2.0*t|.
> I think that's debatable. It worked before and people write it all the
> time. We also write `2*3.141` and it works. I think allowing this is
> reasonable.

Wait, I am confused now. Currently (without your patch), 2*Tensor<1,double> will convert 2 to a double and return a Tensor<1,double>. Does the new version do the same or not?

> In other words, they are identical.

Thank you!

bangerth commented:
> Wait, I am confused now. Currently (without your patch), 2*Tensor will convert 2 to a double and return a Tensor. Does the new version do the same or not?

Yes, it does. I've just added a test for that as well ( #543 ).
